messenger/multiple-buses.rst

```diff
@@ -0,0 +1,261 @@
+.. index::
+    single: Messenger; Multiple buses
+
+Multiple Buses
+==============
+
+A common architecture when building application is to separate commands from
+queries. Commands are actions that do something and queries fetches data. This
+is called CQRS (Command Query Responsibility Segregation). See Martin Fowler's
+`article about CQRS`_ to learn more. This architecture could be used together
+with the messenger component by defining multiple buses.
+
+A **command bus** is a little different form a **query bus**. As example, command
+buses usually don't provide any result and query buses are rarely asynchronous.
+You can configure these buses and their rules by using middleware.
+
+It might also be a good idea to separate actions from reactions by introducing
+an **event bus**. The event bus could have zero or more subscribers.
+
+.. configuration-block::
+
+    .. code-block:: yaml
+
+        framework:
+            messenger:
+                # The bus that is going to be injected when injecting MessageBusInterface
+                default_bus: messenger.bus.commands
+                buses:
+                    messenger.bus.commands:
+                        middleware:
+                            - validation
+                            - doctrine_transaction
+                    messenger.bus.queries:
+                        middleware:
+                            - validation
+                    messenger.bus.events:
+                        middleware:
+                            - allow_no_handler
```

Comment:
Note for 4.2: to be changed in favor of `default_middleware: 'allow_no_handers'`

(cf https://github.com/symfony/symfony/blob/master/UPGRADE-4.2.md#frameworkbundle)